crates/wysiwyg/src/dom/parser/parse.rs

```diff
@@ -12,256 +12,513 @@
-use crate::dom::parser::PaNodeContainer;
 use crate::dom::{Dom, DomCreationError, UnicodeString};
-use crate::ListType;
-
-use super::padom_node::PaDomNode;
-use super::{PaDom, PaDomCreationError, PaDomCreator};
 
 pub fn parse<S>(html: &str) -> Result<Dom<S>, DomCreationError<S>>
 where
     S: UnicodeString,
 {
-    PaDomCreator::parse(html)
-        .map(padom_to_dom)
-        .map_err(padom_creation_error_to_dom_creation_error)
-}
-
-/// Convert a [PaDom] into a [Dom].
-///
-/// [PaDom] is purely used within the parsing process (using html5ever) - in it,
-/// parents refer to their children by handles, and all the nodes are owned in
-/// a big list held by the PaDom itself. PaDoms may also contain garbage nodes
-/// that were created during parsing but are no longer needed. A garbage
-/// collection method was written for testing and is inside padom_creator's
-/// test code. The conversion process here ignores garbage nodes, so they do
-/// not appear in the final Dom.
-///
-/// [Dom] is for general use. Parent nodes own their children, and Dom may be
-/// cloned, compared, and converted into an HTML string.
-fn padom_to_dom<S>(padom: PaDom) -> Dom<S>
-where
-    S: UnicodeString,
-{
-    /// Recurse into panode's children and convert them too
-    fn convert_children<S>(
-        padom: &PaDom,
-        child: &PaNodeContainer,
-        new_node: Option<&mut DomNode<S>>,
-    ) where
-        S: UnicodeString,
-    {
-        if let DomNode::Container(new_node) = new_node.unwrap() {
-            convert(padom, child, new_node);
+    cfg_if::cfg_if! {
+        if #[cfg(feature = "sys")] {
+            sys::parse(html)
+        } else if #[cfg(feature = "js")] {
+            js::parse(html)
         } else {
-            panic!("Container became non-container!");
+            unreachable!("The `sys` or `js` are mutually exclusive, and one of them must be enabled.")
         }
     }
+}
+
+#[cfg(feature = "sys")]
+mod sys {
+    use super::super::padom_node::PaDomNode;
+    use super::super::PaNodeContainer;
+    use super::super::{PaDom, PaDomCreationError, PaDomCreator};
+    use super::*;
+    use crate::dom::nodes::{ContainerNode, DomNode};
+    use crate::ListType;
 
-    /// Create a formatting node
-    fn new_formatting<S>(tag: &str) -> DomNode<S>
+    pub(super) fn parse<S>(html: &str) -> Result<Dom<S>, DomCreationError<S>>
     where
         S: UnicodeString,
     {
-        DomNode::Container(
-            ContainerNode::new_formatting_from_tag(tag.into(), Vec::new())
-                .unwrap_or_else(|| panic!("Unknown format tag {}", tag)),
-        )
+        PaDomCreator::parse(html)
+            .map(padom_to_dom)
+            .map_err(padom_creation_error_to_dom_creation_error)
     }
 
-    /// Create a br node
-    fn new_line_break<S>() -> DomNode<S>
+    /// Convert a [PaDom] into a [Dom].
+    ///
+    /// [PaDom] is purely used within the parsing process (using html5ever) - in it,
```

Comment:
By the way, this whole PaDom structure - if you can see a better way to do it, I'd be really interested!

Reply:
Let me address that in another PR maybe :-).